server/data/data-postgres/src/main/java/org/apache/james/sieve/postgres/PostgresSieveRepository.java

```diff
@@ -100,166 +80,154 @@ private boolean overQuotaAfterModification(long usedSpace, long size, QuotaSizeL
     }
 
     @Override
-    public void putScript(Username username, ScriptName name, ScriptContent content) {
-        transactionRunner.runAndHandleException(Throwing.<EntityManager>consumer(entityManager -> {
-            try {
-                haveSpace(username, name, content.length());
-                JPASieveScript jpaSieveScript = JPASieveScript.builder()
-                        .username(username.asString())
-                        .scriptName(name.getValue())
-                        .scriptContent(content)
-                        .build();
-                entityManager.persist(jpaSieveScript);
-            } catch (QuotaExceededException | StorageException e) {
-                rollbackTransactionIfActive(entityManager.getTransaction());
-                throw e;
-            }
-        }).sneakyThrow(), throwStorageExceptionConsumer("Unable to put script for user " + username.asString()));
+    public void putScript(Username username, ScriptName name, ScriptContent content) throws QuotaExceededException {
+        reThrowQuotaExceededException(() -> spaceThatWillBeUsedByNewScript(username, name, content.length())
+            .flatMap(spaceToUse -> throwOnOverQuota(username, spaceToUse)
+                .thenEmpty(Flux.merge(
+                        updateSpaceUsed(username, spaceToUse),
+                        postgresSieveScriptDAO.upsertScript(PostgresSieveScript.builder()
```

Comment:
update space after inserting the script, using feedback of the insertion (`RETURNING` keyword)?

Reply:
On that, I will tend to rely on the number of inserted/updated rows which is always returned by Postgres to know if the upsert succeeded. RETURNING requires extra data transfer while we do not really need them in this case.